### PR TITLE
Add new label to E0283

### LINKED ISSUE #98938: Calling a trait method as a free function yields odd diagnostics
Given the following code: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=add44602b791902bc38a5e3faf486c45)

```rust
trait Foo {
    fn bar() {}
}

fn main() {
    Foo::bar();
}
```

The current output is:

```
error[E0283]: type annotations needed
 --> src/main.rs:4:5
  |
4 |     Foo::bar();
  |     ^^^^^^^^ cannot infer type
  |
  = note: cannot satisfy `_: Foo`
```

Ideally the output should look like:

```
Cannot call trait method as a free function. It needs an actual type to be implemented on for it to be called.
```

For bonus points: The trait is currently not callable since it's not implemented for anything. It needs to actually be implemented. "It looks like you're trying to call this trait method, but nothing implements it."

### PR DESCRIPTION
**Pull request by obeis**
Closes #98938

### CODE REVIEW

**Comment by compiler-errors** [comment]
Hi @obeis, thanks for the PR. Can you explain the approach you took here a bit more? Specifically:

1. How are we making sure to not say `"cannot call trait method as a free function"` on an expression like `<_ as Trait>::method()`?
1. What's the purpose of restricting to `ObligationCauseCode::ItemObligation`? Does this suppress erroneous suggestions you were seeing while testing?
1. Why are you checking `param_env.is_const()`? Same as above, does this suppress erroneous suggestions?

Please also provide a [UI test](https://rustc-dev-guide.rust-lang.org/tests/adding.html) or a few demonstrating your changes.

**Comment by compiler-errors** [comment]
```
LL |     let mut _q = Default::default();
   |                  ^^^^^^^^^^^^^^^^ cannot call trait method as a free function
```

The message is a bit confusing in this context. The problem here is not that we can't call `Default::default` as a free method, but instead that we're not able to infer the _return type_ of `Default::default`. 

It we are going to suggest this message, we may need to restrict it to function signatures that do not contain the ambiguous inference variable in either its inputs or outputs.

**Comment by obeis** [comment]
Hi @compiler-errors, thanks for taking time to review this.
- On an expression like `<_ as Trait>::method()` the error emit with the new label so this need a fix.
- Restricting to `ObligationCauseCode::ItemObligation` because this suppress erroneous suggestions i was seeing while testing.
- Checking `param_env.is_const()` to suppress erroneous suggestions, like:
```
error[E0283]: type annotations needed
  --> /home/obei/rust/rust/src/test/ui/associated-consts/issue-63496.rs:4:21
   |
LL |     fn f() -> ([u8; A::C], [u8; A::C]);
   |                     ^^^^ cannot call trait method as a free function

error[E0283]: type annotations needed
  --> /home/obei/rust/rust/src/test/ui/associated-consts/issue-63496.rs:4:33
   |
LL |     fn f() -> ([u8; A::C], [u8; A::C]);
   |                                 ^^^^ cannot call trait method as a free function

```
ui test [code](https://github.com/rust-lang/rust/blob/d695a497bbf4b20d2580b75075faa80230d41667/src/test/ui/associated-consts/issue-63496.rs), [stderr](https://github.com/rust-lang/rust/blob/d695a497bbf4b20d2580b75075faa80230d41667/src/test/ui/associated-consts/issue-63496.stderr) for code

**Comment by obeis** [comment]
@rustbot ready
@compiler-errors Please have a look

File: src/test/ui/error-codes/E0283.stderr [hunk 1]
```diff
@@ -5,31 +5,13 @@ LL |     fn create() -> u32;
...
-LL | impl Into<u32> for Impl {
-   | ^^^^^^^^^^^^^^^^^^^^^^^
-   = note: and another `impl` found in the `core` crate:
-           - impl<T, U> Into<U> for T
-             where U: From<T>;
-help: try using a fully qualified path to specify the expected types
-   |
-LL |     let bar = <Impl as Into<T>>::into(foo_impl) * 1u32;
-   |               ++++++++++++++++++++++++        ~
+   |                        ^^^^ cannot call trait method as a free function
```
Comment (cjgillot) [inline src/test/ui/error-codes/E0283.stderr:14]: This note is wrong.

**Pull request closed without merging**